File: ELProgressHUD/SBProgressHud/ELProgressHud.m
//
//  ELProgressHud.m
//  ELProgressHud
//
//  Created by 李涌辉 on 16/8/27.
//  Copyright © 2016年 LiYonghui~. All rights reserved.
//

#import "ELProgressHud.h"

CGFloat width = 160;
CGFloat height = 120;
CGFloat const ELProgressHudTag = 962;
@interface ELProgressHud ()

@property (nonatomic, strong) UIView *mainView;
@property (nonatomic, strong) UIActivityIndicatorView *ELActivityIndicatorView;

@property (nonatomic, strong) UIProgressView *ELProgressView;
@property (nonatomic, strong) UILabel *ELProgressLabel;
@property (nonatomic, strong) UIButton *ELCancelButton;
@property (nonatomic, strong) UIImageView *ELImageView;
@property (nonatomic, strong) CAShapeLayer *ELCircleShapeLayer;
@property (nonatomic, strong) CAShapeLayer *ELProgressShapeLayer;

@end

@implementation ELProgressHud

#pragma mark - initialize hud

- (instancetype)init {
    self = [super init];
    if (self) {
        _hudType = ELProgressHudTypeIndicator;
        [self setup];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        
    }
    return self;
}

- (instancetype)initWithHudType:(ELProgressHudType)hudType {
    self = [super init];
    if (self) {
        _hudType = hudType;
        [self setup];
    }
    return self;
}

#pragma mark - setup

- (void)setup {
    self.frame = CGRectMake(0, 0, CGRectGetWidth([UIScreen mainScreen].bounds), CGRectGetHeight([UIScreen mainScreen].bounds));
    self.mainView.frame = CGRectMake(0, 0, width, height);
    self.mainView.center = CGPointMake(CGRectGetMidX([UIScreen mainScreen].bounds), CGRectGetMidY([UIScreen mainScreen].bounds) - 20);
    [self addSubview:self.mainView];
    switch (_hudType) {
        case ELProgressHudTypeIndicator:
        {
            width = 80;
            height = 80;
            self.ELActivityIndicatorView.frame = CGRectMake(0, 0, 50, 50);
            self.ELActivityIndicatorView.center = CGPointMake(width / 2, height / 2);
            [self.mainView addSubview:self.ELActivityIndicatorView];
        }
            break;
        case ELProgressHudTypeText:
        {
            height = 60;
            self.textLabel.frame = CGRectMake(0, 0, width, 44);
            self.textLabel.center = CGPointMake(width / 2, height / 2);
            [self.mainView addSubview:self.textLabel];
        }
            break;
        case ELProgressHudTypeProgress:
        {
            height = 80;
            [self setupELProgressLabel];
            [self setupELProgressView];
            [self.mainView addSubview:self.ELProgressLabel];
            [self.mainView addSubview:self.ELProgressView];
        }
            break;
        case ELProgressHudTypeProgressWithCancel:
        {
            [self setupELProgressLabel];
            [self setupELProgressView];
            self.ELCancelButton.frame = CGRectMake(0, 0, width * 0.4, 36);
            self.ELCancelButton.center = CGPointMake(width / 2, height * 0.8);
            [self.mainView addSubview:self.ELProgressLabel];
            [self.mainView addSubview:self.ELProgressView];
            [self.mainView addSubview:self.ELCancelButton];
        }
            break;
        case ELProgressHudTypeIndicatorAndText:
        {
            [self setupELActivityIndicatorView];
            [self setupTextLabel];
            [self.mainView addSubview:self.ELActivityIndicatorView];
            [self.mainView addSubview:self.textLabel];
            
        }
            break;
        case ELProgressHudTypeProgressAndText:
        {
            [self setupELProgressLabel];
            [self setupELProgressView];
            [self setupTextLabel];
            [self.mainView addSubview:self.ELProgressLabel];
            [self.mainView addSubview:self.ELProgressView];
            [self.mainView addSubview:self.textLabel];
        }
            break;
        case ELProgressHudTypeGif:
        {
            self.ELImageView.frame = self.mainView.bounds;
            [self.mainView addSubview:self.ELImageView];
        }
            break;
        case ELProgressHudTypeDrawCircle:
        {
            width = 120;
            self.ELProgressShapeLayer.frame = self.mainView.bounds;
            self.ELCircleShapeLayer.frame = self.mainView.bounds;
            [self.mainView.layer addSublayer:self.ELCircleShapeLayer];
            [self.mainView.layer addSublayer:self.ELProgressShapeLayer];
            self.ELProgressLabel.frame = CGRectMake(0, 0, width * 0.5, 30);
            self.ELProgressLabel.center = CGPointMake(width / 2, height / 2);
            [self.mainView addSubview:self.ELProgressLabel];
        }
            break;
        default:
            break;
    }
}

- (void)layoutSubviews {
    self.mainView.frame = CGRectMake(0, 0, width, height);
    self.mainView.center = CGPointMake(CGRectGetMidX([UIScreen mainScreen].bounds), CGRectGetMidY([UIScreen mainScreen].bounds) - 20);
}

- (void)setupELActivityIndicatorView {
    self.ELActivityIndicatorView.frame = CGRectMake(0, 0, 50, 50);
    self.ELActivityIndicatorView.center = CGPointMake(width / 2, 50);
}

- (void)setupELProgressLabel {
    self.ELProgressLabel.frame = CGRectMake(0, 0, width * 0.5, 30);
    self.ELProgressLabel.center = CGPointMake(width / 2, height * 0.2);
}

- (void)setupELProgressView {
    self.ELProgressView.frame = CGRectMake(0, 0, width * 0.8, 1);
    self.ELProgressView.center = CGPointMake(width / 2, height * 0.5);
    self.ELProgressView.transform = CGAffineTransformMakeScale(1.0f,3.0f);
}

- (void)setupTextLabel {
    self.textLabel.frame = CGRectMake(0, 0, width, 44);
    self.textLabel.center = CGPointMake(width / 2, height - 30);
}

#pragma mark - public methods

+ (void)showHudAtView:(UIView *)aView {
    ELProgressHud *hud = [[ELProgressHud alloc] init];
    hud.tag = ELProgressHudTag;
    [hud.ELActivityIndicatorView startAnimating];
    [aView addSubview:hud];
}

+ (void)hideHudAtView:(UIView *)aView {
    ELProgressHud *hud = [aView viewWithTag:ELProgressHudTag];
    if (hud) {
        hud.mainView = nil;
        hud.ELActivityIndicatorView = nil;
        width = 160;
        height = 120;
        [hud removeFromSuperview];
        hud = nil;
    }
}

- (void)showHudAtView:(UIView *)aView {
    self.tag = ELProgressHudTag;
    [self.ELActivityIndicatorView startAnimating];
    [aView addSubview:self];
    if (self.delay) {
        [self performSelector:@selector(hide) withObject:nil afterDelay:self.delay];
    }
}

#pragma mark - private method

- (void)hide {
    self.mainView = nil;
    self.ELActivityIndicatorView = nil;
    width = 160;
    height = 120;
    [self removeFromSuperview];
}

#pragma mark - cancelbutton event

- (void)ELCancelButtonClick {
    if (self.elCompletion) {
        self.elCompletion();
    }
    self.elCompletion = nil;
    [self hide];
}

#pragma mark - setter & getter

- (UIView *)mainView {
    if (!_mainView) {
        _mainView = [UIView new];
        _mainView.backgroundColor = [UIColor lightGrayColor];
        _mainView.layer.cornerRadius = 8;
        _mainView.layer.masksToBounds = YES;
    }
    return _mainView;
}

- (UIActivityIndicatorView *)ELActivityIndicatorView {
    if (!_ELActivityIndicatorView) {
        _ELActivityIndicatorView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
    }
    return _ELActivityIndicatorView;
}

- (UILabel *)textLabel {
    if (!_textLabel) {
        _textLabel = [[UILabel alloc] init];
        _textLabel.text = @"Loading";
        _textLabel.textAlignment = NSTextAlignmentCenter;
    }
    return _textLabel;
}

- (UIProgressView *)ELProgressView {
    if (!_ELProgressView) {
        _ELProgressView = [[UIProgressView alloc] init];
        _ELProgressView.progressTintColor = [UIColor orangeColor];
    }
    return _ELProgressView;
}

- (void)setEl_progress:(float)el_progress {
    if ([NSThread isMainThread]) {
        [self.ELProgressView setProgress:el_progress animated:YES];
        [self.ELProgressLabel setText:[NSString stringWithFormat:@"%.2f%%", el_progress * 100]];
    }else {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.ELProgressView setProgress:el_progress animated:YES];
            [self.ELProgressLabel setText:[NSString stringWithFormat:@"%.2f%%", el_progress * 100]];
        });
    }
}

- (UILabel *)ELProgressLabel {
    if (!_ELProgressLabel) {
        _ELProgressLabel = [[UILabel alloc] init];
        _ELProgressLabel.textAlignment = NSTextAlignmentCenter;
        _ELProgressLabel.textColor = [UIColor lightTextColor];
    }
    return _ELProgressLabel;
}

- (UIButton *)ELCancelButton {
    if (!_ELCancelButton) {
        _ELCancelButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_ELCancelButton addTarget:self action:@selector(ELCancelButtonClick) forControlEvents:UIControlEventTouchUpInside];
        _ELCancelButton.backgroundColor = [UIColor grayColor];
        [_ELCancelButton setTitle:@"Cancel" forState:UIControlStateNormal];
    }
    return _ELCancelButton;
}

- (UIImageView *)ELImageView {
    if (!_ELImageView) {
        _ELImageView = [[UIImageView alloc] init];
        _ELImageView.alpha = 0.5;
    }
    return _ELImageView;
}

- (void)setEl_images:(NSArray<NSString *> *)el_images {
    NSMutableArray *imageArray = [NSMutableArray array];
    for (NSString *imageName in el_images) {
        UIImage *image = [UIImage imageNamed:imageName];
        [imageArray addObject:image];
    }
    self.ELImageView.animationImages = imageArray;
    [self.ELImageView startAnimating];
}

- (CAShapeLayer *)ELCircleShapeLayer {
    if (!_ELCircleShapeLayer) {
        UIBezierPath *path = [UIBezierPath bezierPathWithOvalInRect:CGRectMake(width * (1 - 0.6) / 2, height * (1 - 0.6) / 2, width * 0.6, height * 0.6)];
        _ELCircleShapeLayer = [CAShapeLayer layer];
        _ELCircleShapeLayer.fillColor = [UIColor clearColor].CGColor;
        _ELCircleShapeLayer.strokeColor=[UIColor orangeColor].CGColor;
        _ELCircleShapeLayer.path = path.CGPath;
        _ELCircleShapeLayer.lineWidth = 3;
    }
    return _ELCircleShapeLayer;
}

- (CAShapeLayer *)ELProgressShapeLayer {
    if (!_ELProgressShapeLayer) {
        UIBezierPath *path = [UIBezierPath bezierPathWithOvalInRect:CGRectMake(width * (1 - 0.6) / 2, height * (1 - 0.6) / 2, width * 0.6, height * 0.6)];
        _ELProgressShapeLayer = [CAShapeLayer layer];
        _ELProgressShapeLayer.fillColor = [UIColor clearColor].CGColor;
        _ELProgressShapeLayer.strokeColor=[UIColor whiteColor].CGColor;
        _ELProgressShapeLayer.path = path.CGPath;
        _ELProgressShapeLayer.lineWidth = 3;
    }
    return _ELProgressShapeLayer;
}

- (void)setEl_strokeStart:(CGFloat)el_strokeStart {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.ELProgressShapeLayer.strokeStart = el_strokeStart;
        [self.ELProgressLabel setText:[NSString stringWithFormat:@"%.2f%%", el_strokeStart * 100]];
    });
}

@end

File: ELProgressHUD-swift/ELProgressHud.swift
//
//  ELProgressHud.swift
//  ELProgressHUD-swift
//
//  Created by 李涌辉 on 16/9/18.
//  Copyright © 2016年 Evil Lee. All rights reserved.
//

import UIKit

private var width : CGFloat = 160
private var height : CGFloat = 120
private let ELProgressHudTag : NSInteger = 962

class ELProgressHud: UIView {
    
    public enum ELProgressHudType : NSInteger {
        case ELProgressHudTypeIndicator = 0// 普通菊花
        case ELProgressHudTypeText
        case ELProgressHudTypeProgress
        case ELProgressHudTypeProgressWithCancel
        case ELProgressHudTypeIndicatorAndText
        case ELProgressHudTypeProgressAndText
        case ELProgressHudTypeGif
        case ELProgressHudTypeDrawCircle
    }
    
    typealias ELCompletion = () -> ()
    var hudType : ELProgressHudType = .ELProgressHudTypeIndicator
    var delay : TimeInterval?
    lazy var textLabel : UILabel = {
        let textLabel = UILabel()
        textLabel.text = "Loading"
        textLabel.textAlignment = .center
        return textLabel
    }()
    
    var el_progress : Float {
        get {
          return self.el_progress
        }
        set(progress) {
        self.ELProgressView.progress = progress
        }
    }
    
    var elCompletion : ELCompletion?
    var el_images : [String] {
        get {
         return self.el_images
        }
        set(images) {
            var imageArray: [UIImage] = [UIImage].init()
            for imageName: String in images {
                let image: UIImage = UIImage.init(named: imageName)!
                imageArray.append(image)
            }
            self.ELImageView.animationImages = imageArray
            self.ELImageView.startAnimating()
        }
    }
    var el_strokeStart : CGFloat {
        get {
            return self.el_strokeStart
        }
        set(strokeStart) {
           self.ELProgressShapeLayer.strokeStart = strokeStart
        }
    }
    
    
    lazy private var mainView : UIView? = {
        let mainView = UIView()
        mainView.backgroundColor = UIColor.lightGray
        mainView.layer.cornerRadius = 8.0
        mainView.layer.masksToBounds = true
        return mainView
    }()

    lazy private var ELActivityIndicatorView : UIActivityIndicatorView? = {
        let ELActivityIndicatorView = UIActivityIndicatorView.init(activityIndicatorStyle: .whiteLarge)
        return ELActivityIndicatorView
    }()
    
    lazy private var ELProgressView : UIProgressView = {
        let ELProgressView = UIProgressView()
        ELProgressView.progressTintColor = UIColor.orange
        return ELProgressView
    }()
    
    lazy private var ELProgressLabel : UILabel = {
        let ELProgressLabel = UILabel()
        ELProgressLabel.textAlignment = .center
        ELProgressLabel.textColor = UIColor.lightText
        return ELProgressLabel
    }()
    
    lazy private var ELCancelButton : UIButton = {
        let ELCancelButton = UIButton.init(type: .custom)
        ELCancelButton.addTarget(self, action: #selector(ELCancelButtonClick), for: .touchUpInside)
        ELCancelButton.backgroundColor = UIColor.gray
        ELCancelButton.setTitle("Cancel", for: .normal)
        return ELCancelButton
    }()
    
    lazy private var ELImageView : UIImageView = {
        let ELImageView = UIImageView()
        ELImageView.alpha = 0.5
        return ELImageView
    }()
    
    lazy private var ELCircleShapeLayer : CAShapeLayer = {
        let path : UIBezierPath = UIBezierPath.init(ovalIn: CGRect.init(x: width * (1 - 0.6) / 2, y: height * (1 - 0.6) / 2, width: width * 0.6, height: height * 0.6))
        let ELCircleShapeLayer = CAShapeLayer()
        ELCircleShapeLayer.fillColor = UIColor.clear.cgColor
        ELCircleShapeLayer.strokeColor = UIColor.orange.cgColor
        ELCircleShapeLayer.path = path.cgPath
        ELCircleShapeLayer.lineWidth = 3
        return ELCircleShapeLayer
    }()
    
    lazy private var ELProgressShapeLayer : CAShapeLayer = {
        let path : UIBezierPath = UIBezierPath.init(ovalIn: CGRect.init(x: width * (1 - 0.6) / 2, y: height * (1 - 0.6) / 2, width: width * 0.6, height: height * 0.6))
        let ELProgressShapeLayer = CAShapeLayer()
        ELProgressShapeLayer.fillColor = UIColor.clear.cgColor
        ELProgressShapeLayer.strokeColor = UIColor.white.cgColor
        ELProgressShapeLayer.path = path.cgPath
        ELProgressShapeLayer.lineWidth = 3
        return ELProgressShapeLayer
    }()
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    

    required init(hudType: ELProgressHudType) {
        super.init(frame: UIScreen.main.bounds)
        self.hudType = hudType
        setup()
    }
    required override init(frame: CGRect) {
        super.init(frame: frame)
        self.hudType = .ELProgressHudTypeIndicator
        setup()
    }
    
    func showHudAtView(aView : UIView) -> Void {
        self.tag = ELProgressHudTag
        self.ELActivityIndicatorView!.startAnimating()
        aView.addSubview(self)
        if self.delay != 0 && self.delay != nil {
            self.perform(#selector(hide), with: nil, afterDelay: self.delay!)
        }
    }
    
    class func showHudAtView(aView : UIView) -> Void {
        let hud : ELProgressHud = ELProgressHud.init(frame: UIScreen.main.bounds)
        hud.tag = ELProgressHudTag
        hud.ELActivityIndicatorView!.startAnimating()
        aView.addSubview(hud)
    }
    
    class func hideHudAtView(aView : UIView) -> Void {
        var hud : ELProgressHud? = aView.viewWithTag(ELProgressHudTag) as? ELProgressHud
        if hud !== nil {
            hud?.mainView = nil
            hud?.ELActivityIndicatorView = nil
            width = 160
            height = 120
            hud?.removeFromSuperview()
            hud = nil
        }
    }
    
    private func setup() {
      self.frame = CGRect.init(x: 0, y: 0, width: UIScreen.main.bounds.width, height: UIScreen.main.bounds.height)
        self.mainView?.frame = CGRect.init(x: 0, y: 0, width: width, height: height)
        self.mainView?.center = CGPoint.init(x: UIScreen.main.bounds.midX, y: UIScreen.main.bounds.midY - 20)
        self.addSubview(self.mainView!)
        switch self.hudType {
        case .ELProgressHudTypeIndicator:
                width = 80
                height = 80
                self.ELActivityIndicatorView?.frame = CGRect.init(x: 0, y: 0, width: 50, height: 50)
                self.ELActivityIndicatorView?.center = CGPoint.init(x: width / 2, y: height / 2)
                self.mainView?.addSubview(self.ELActivityIndicatorView!)
            
        case .ELProgressHudTypeText:
            height = 60
            self.textLabel.frame = CGRect.init(x: 0, y: 0, width: width, height: 44)
            self.textLabel.center = CGPoint.init(x: width / 2, y: height / 2)
            self.mainView?.addSubview(self.textLabel)
            
        case .ELProgressHudTypeProgress:
            height = 80
            setupELProgressLabel()
            setupELProgressView()
            self.mainView?.addSubview(self.ELProgressLabel)
            self.mainView?.addSubview(self.ELProgressView)
            
        case .ELProgressHudTypeProgressWithCancel:
            setupELProgressLabel()
            setupELProgressView()
            self.ELCancelButton.frame = CGRect.init(x: 0, y: 0, width: width * 0.4, height: 36)
            self.ELCancelButton.center = CGPoint.init(x: width / 2, y: height * 0.8)
            self.mainView?.addSubview(self.ELProgressLabel)
            self.mainView?.addSubview(self.ELProgressView)
            self.mainView?.addSubview(self.ELCancelButton)
            
        case .ELProgressHudTypeIndicatorAndText:
            setupELActivityIndicatorView()
            setupTextLabel()
            self.mainView?.addSubview(self.ELActivityIndicatorView!)
            self.mainView?.addSubview(self.textLabel)
            
        case .ELProgressHudTypeProgressAndText:
            setupELProgressLabel()
            setupELProgressView()
            setupTextLabel()
            self.mainView?.addSubview(self.ELProgressLabel)
            self.mainView?.addSubview(self.ELProgressView)
            self.mainView?.addSubview(self.textLabel)
            
        case .ELProgressHudTypeGif:
            self.ELImageView.frame = (self.mainView?.bounds)!
            self.mainView?.addSubview(self.ELImageView)
            
        case .ELProgressHudTypeDrawCircle:
            width = 120
            self.ELProgressShapeLayer.frame = (self.mainView?.bounds)!
            self.ELCircleShapeLayer.frame = (self.mainView?.bounds)!
            self.mainView?.layer.addSublayer(self.ELCircleShapeLayer)
            self.mainView?.layer.addSublayer(self.ELProgressShapeLayer)
            self.ELProgressLabel.frame = CGRect.init(x: 0, y: 0, width: width * 0.5, height: 30)
            self.ELProgressLabel.center = CGPoint.init(x: width / 2, y: height / 2)
            self.mainView?.addSubview(self.ELProgressLabel)
            
        }
    }
    
    override func layoutSubviews() {
        self.mainView?.frame = CGRect.init(x: 0, y: 0, width: width, height: height)
        self.mainView?.center = CGPoint.init(x: UIScreen.main.bounds.midX, y: UIScreen.main.bounds.midY - 20)
    }
    
    private func setupELActivityIndicatorView() {
        self.ELActivityIndicatorView?.frame = CGRect.init(x: 0, y: 0, width: 50, height: 50)
        self.ELActivityIndicatorView?.center = CGPoint.init(x: width / 2, y: 50)
    }
    
    private func setupELProgressLabel() {
        self.ELProgressLabel.frame = CGRect.init(x: 0, y: 0, width: width / 2, height: 30)
        self.ELProgressLabel.center = CGPoint.init(x: width / 2, y: height * 0.2)
    }
    
    private func setupELProgressView() {
        self.ELProgressView.frame = CGRect.init(x: 0, y: 0, width: width * 0.8, height: 1)
        self.ELProgressView.center = CGPoint.init(x: width / 2, y: height / 2)
        //self.ELProgressView.transform = cgaffinetransform
    }
    
    private func setupTextLabel() {
        self.textLabel.frame = CGRect.init(x: 0, y: 0, width: width, height: 44)
        self.textLabel.center = CGPoint.init(x: width / 2, y: height - 30)
    }
    
    @objc private func hide() {
        self.mainView = nil
        self.ELActivityIndicatorView = nil
        width = 160
        height = 120
        self.removeFromSuperview()
    }
    
    @objc private func ELCancelButtonClick() {
        if self.elCompletion != nil {
            self.elCompletion!()
        }
        self.elCompletion = nil
        self.hide()
    }
    
}
